packages/utils/telemetry-utils/src/errorLogging.ts

```diff
@@ -3,102 +3,202 @@
+ * @returns A new error object "wrapping" the given error
+ */
+export function wrapError<T>(
+    error: any,
+    newErrorFn: (m: string) => T,
+): T {
+    const {
         message,
-    };
+        stack,
+    } = extractLogSafeErrorProperties(error);
+    const props = isILoggingError(error) ? error.getTelemetryProperties() : {};
 
-    if (isRegularObject(error)) {
-        const { errorType, stack, name } = error;
+    const newError = newErrorFn(message);
+    mixinTelemetryProps(newError, props);
 
-        if (typeof errorType === "string") {
-            safeProps.errorType = errorType;
-        }
-
-        if (typeof stack === "string") {
-            const errorName = (typeof name === "string") ? name : undefined;
-            safeProps.stack = removeMessageFromStack(stack, errorName);
-        }
+    if (stack !== undefined) {
+        Object.assign(newError, { stack });
     }
 
-    return safeProps;
+    return newError;
 }
 
-/** Type guard for ILoggingError */
-export const isILoggingError = (x: any): x is ILoggingError => typeof x?.getTelemetryProperties === "function";
+/**
+ * Normalize the given error yielding a valid Fluid Error
+ * @returns The same error object passed in if possible, normalized and with any provided annotations applied
+ * @param error - The error to normalize, ideally by patching in the properties of IFluidErrorBase
+ * @param annotations - Annotations to apply to the normalized error
+ */
+export function normalizeError(
+    error: unknown,
+    annotations: FluidErrorAnnotations = {},
+): IFluidErrorBase {
+    if (isRegularObject(error) && !Object.isFrozen(error)) {
+        // We can simply annotate the error and return it
+        annotateErrorObject(error, annotations);
+        return error;
+    }
+
+    // We can't annotate the error, so we have to wrap it
+    const newErrorFn = (errMsg: string) => new LoggingError(errMsg);
+    const errorObject = wrapError<LoggingError>(error, newErrorFn);
+    const errorTypeIfNone = !isRegularObject(error) ? typeof(error) : "wrappedFrozenError";
+    const fluidError = patchFluidErrorBuilder(
+        errorObject as FluidErrorBuilder,
+        prepareFluidErrorTemplate(errorObject, errorTypeIfNone, annotations.errorCodeIfNone),
+    );
+
+    mixinTelemetryPropsWithFluidError(fluidError, annotations.props);
+    return fluidError;
+}
 
 /**
- * Read-Write Logging Error.  Not exported.
- * This type alias includes addTelemetryProperties, and applies to objects even if not instanceof LoggingError\
+ * Annotate the given error object, such that it becomes a valid Fluid Error (and is asserted so via a type guard)
+ * If the object is not suitable for annotation (e.g. it's frozen), this will throw.
+ * @param error - The error to annotate by patching in the properties of IFluidErrorBase
+ * @param annotations - Annotations to apply to the normalized error
  */
-type RwLoggingError = LoggingError;
+export function annotateErrorObject(
```

Comment:
I'd think that this function should be called only for error objects of specific type, i.e. error object should have error code. If not - it should fail, and instead, normalizeError() should be used, no?

Reply:
Basically I'd expect it to be called to only add properties, and not deal with error codes. And the other function to be more generic in terms of creating of object of the right structure/type and thus require error code. Right now it's optional and is on both paths, which makes it fuzzy for me on when to call either of the functions.

Reply:
I see what you're saying.

Maybe we could constrain this to IFluidErrorBase later, but while that's still rolling out (first introduced in this PR!), I want to have onramps to get stuff lined up with the new interface.

Then it will be an iterative process of looking at telemetry data where error code starts with "none", and pushing for instrumenting with error code earlier and earlier.

The idea of errorCodeIfNone in both functions is to provide a "where" hint for errors we didn't raise ourselves (or did but haven't upgraded to IFluidErrorBase yet).

There are two that provide different syntax opportunities.  With annotateErrorObject you don't assign the result to something, you just get type narrowing on the passed in error. NormalizeError makes no assumptions about the input and you should assign the result to a new variable (even if it's the same object as was provided).

Reply:
From usage POV, I believe you have to use annotateErrorObject() only for known code paths (i.e. caller has to know for sure that error was generated by our code and has proper structure). If that's true, then I'd rather encode it in the signature. If it's not true, then I do not understand what is the guidance - what should I use annotateErrorObject vs. NormalizeError ? I'm more focused on message to developers, less about process on what happens next (looking at telemetry).

Reply:
I'm considering making this tighter in a subsequent PR